File: lib/screens/home/views/home_screen.dart
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:weather_app/provider/home_provider.dart';

class HomeScreen extends StatefulWidget {
  const HomeScreen({super.key});

  @override
  State<HomeScreen> createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  TextEditingController searchController = TextEditingController();

  @override
  void initState() {
    super.initState();
    context.read<HomeProvider>().loadBookmarkedCities();
  }

  void searchCityWeather() {
    String city = searchController.text.trim();
    if (city.isNotEmpty) {
      context.read<HomeProvider>().fetchWeatherData(city: city);
    }
  }

  void addToBookmark(String city) {
    context.read<HomeProvider>().bookmarkCity(city).then((_) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(
          content: Text('$city has been added to bookmarks'),
          backgroundColor: Colors.green,
        ),
      );
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Consumer<HomeProvider>(
        builder: (context, provider, child) {
          if (provider.weatherModel == null) {
            return Center(
              child: Padding(
                padding: const EdgeInsets.all(28.0),
                child: Column(
                  mainAxisAlignment: MainAxisAlignment.center,
                  crossAxisAlignment: CrossAxisAlignment.center,
                  children: [
                    // Welcome message
                    Text(
                      'Welcome to WeatherApp!',
                      style: TextStyle(
                        fontSize: 30,
                        fontWeight: FontWeight.bold,
                        color: Colors.black38,
                        shadows: [
                          Shadow(
                            color: Colors.black.withOpacity(0.5),
                            offset: const Offset(2, 2),
                            blurRadius: 8,
                          ),
                        ],
                      ),
                      textAlign: TextAlign.center,
                    ),
                    const SizedBox(height: 30),

                    const Padding(
                      padding: EdgeInsets.symmetric(horizontal: 40.0),
                      child: Text(
                        'Enter a city name to get the latest weather information.',
                        style: TextStyle(
                          fontSize: 18,
                          color: Colors.white70,
                          fontWeight: FontWeight.w300,
                        ),
                      ),
                    ),
                    const SizedBox(height: 40),

                    Container(
                      padding: const EdgeInsets.all(16.0),
                      decoration: BoxDecoration(
                        color: Colors.white.withOpacity(0.8),
                        borderRadius: BorderRadius.circular(30.0),
                        boxShadow: [
                          BoxShadow(
                            color: Colors.black.withOpacity(0.1),
                            blurRadius: 12,
                            offset: const Offset(0, 4),
                          ),
                        ],
                      ),
                      child: Row(
                        children: [
                          Expanded(
                            child: TextField(
                              controller: searchController,
                              decoration: const InputDecoration(
                                hintText: "Enter city name",
                                border: InputBorder.none,
                                contentPadding: EdgeInsets.all(16.0),
                                hintStyle: TextStyle(
                                  fontSize: 16.0,
                                  color: Colors.grey,
                                ),
                              ),
                              style: const TextStyle(
                                fontSize: 18.0,
                                fontWeight: FontWeight.bold,
                                color: Colors.black,
                              ),
                              onSubmitted: (value) {
                                searchCityWeather();
                              },
                            ),
                          ),
                          IconButton(
                            icon: const Icon(Icons.search, color: Colors.blue),
                            onPressed: () {
                              searchCityWeather();
                            },
                          ),
                        ],
                      ),
                    ),
                  ],
                ),
              ),
            );
          }

          String backgroundImage = provider.getBackgroundImage(
              provider.weatherModel!.weathersList![0].icon!);

          return SingleChildScrollView(
            child: Container(
              decoration: BoxDecoration(
                gradient: LinearGradient(
                  begin: Alignment.topLeft,
                  end: Alignment.bottomRight,
                  colors: [
                    Colors.blue.withOpacity(0.7),
                    Colors.white.withOpacity(0.5)
                  ],
                ),
                image: DecorationImage(
                  image: AssetImage(backgroundImage),
                  fit: BoxFit.cover,
                  opacity: 0.5,
                ),
              ),
              child: Padding(
                padding: const EdgeInsets.all(16.0),
                child: Column(
                  crossAxisAlignment: CrossAxisAlignment.start,
                  children: [
                    Padding(
                      padding: const EdgeInsets.only(top: 16.0),
                      child: Expanded(
                        child: Container(
                          decoration: BoxDecoration(
                            color: Colors.white.withOpacity(0.8),
                            borderRadius: BorderRadius.circular(30.0),
                            boxShadow: const [
                              BoxShadow(
                                color: Colors.blueGrey,
                                offset: Offset(4, 4),
                                blurRadius: 10.0,
                              ),
                            ],
                          ),
                          child: Row(
                            children: [
                              Expanded(
                                child: TextField(
                                  controller: searchController,
                                  decoration: const InputDecoration(
                                    hintText: "Enter city name",
                                    border: InputBorder.none,
                                    contentPadding: EdgeInsets.all(16.0),
                                  ),
                                  onSubmitted: (value) {
                                    searchCityWeather();
                                  },
                                ),
                              ),
                              IconButton(
                                icon: const Icon(Icons.search,
                                    color: Colors.blue),
                                onPressed: () {
                                  searchCityWeather();
                                },
                              ),
                            ],
                          ),
                        ),
                      ),
                    ),
                    const SizedBox(height: 16.0),
                    if (provider.bookmarkedCities.isNotEmpty)
                      Padding(
                        padding: const EdgeInsets.only(bottom: 16.0),
                        child: Row(
                          children: [
                            Container(
                              padding: const EdgeInsets.all(12.0),
                              decoration: BoxDecoration(
                                color: Colors.white.withOpacity(0.5),
                                borderRadius: BorderRadius.circular(30.0),
                                boxShadow: [
                                  BoxShadow(
                                    color: Colors.black.withOpacity(0.2),
                                    blurRadius: 10.0,
                                    offset: const Offset(0, 4),
                                  ),
                                ],
                              ),
                              child: Row(
                                children: [
                                  const Icon(
                                    Icons.bookmark,
                                    color: Colors.amber,
                                    size: 28.0,
                                  ),
                                  const SizedBox(width: 6.0),
                                  Text(
                                    'Bookmarked City: ${provider.bookmarkedCities.last}',
                                    style: const TextStyle(
                                      fontSize: 20,
                                      fontWeight: FontWeight.bold,
                                      color: Colors.blue,
                                    ),
                                  ),
                                ],
                              ),
                            ),
                          ],
                        ),
                      ),
                    if (provider.weatherModel != null)
                      Column(
                        crossAxisAlignment: CrossAxisAlignment.start,
                        children: [
                          Row(
                            mainAxisAlignment: MainAxisAlignment.spaceBetween,
                            children: [
                              Row(
                                crossAxisAlignment: CrossAxisAlignment.start,
                                children: [
                                  Text(
                                    provider.weatherModel!.name ??
                                        "City not found",
                                    style: const TextStyle(
                                      fontSize: 34,
                                      fontWeight: FontWeight.bold,
                                      color: Colors.white,
                                      shadows: [
                                        Shadow(
                                          blurRadius: 5.0,
                                          color: Colors.black26,
                                          offset: Offset(2.0, 2.0),
                                        ),
                                      ],
                                    ),
                                  ),
                                  const SizedBox(width: 5),
                                  Text(
                                    provider.weatherModel?.sysModel?.country ??
                                        "Country not found",
                                    style: const TextStyle(
                                      fontSize: 22,
                                      fontWeight: FontWeight.w500,
                                      color: Colors.white,
                                      letterSpacing: 1.2,
                                    ),
                                  ),
                                ],
                              ),
                              IconButton(
                                icon: const Icon(
                                  Icons.bookmark,
                                  color: Colors.yellow,
                                  size: 30,
                                ),
                                onPressed: () {
                                  String city = searchController.text.trim();
                                  if (city.isNotEmpty) {
                                    addToBookmark(city);
                                  }
                                },
                              ),
                            ],
                          ),
                          const SizedBox(height: 6.0),
                          Align(
                            alignment: Alignment.center,
                            child: Text(
                              provider.weatherList.isNotEmpty
                                  ? provider.weatherList.first.description ?? ""
                                  : "No weather data available",
                              style: const TextStyle(
                                fontSize: 24,
                                fontWeight: FontWeight.w500,
                                color: Colors.white,
                                letterSpacing: 1.0,
                                shadows: [
                                  Shadow(
                                    blurRadius: 3.0,
                                    color: Colors.black26,
                                    offset: Offset(1.0, 1.0),
                                  ),
                                ],
                              ),
                            ),
                          ),
                          const SizedBox(height: 8),
                          Row(
                            mainAxisAlignment: MainAxisAlignment.center,
                            children: [
                              AnimatedRotation(
                                turns: 1,
                                duration: const Duration(seconds: 2),
                                child: Image.asset(
                                  provider.getWeatherIcon(provider
                                      .weatherModel!.weathersList![0].icon!),
                                  width: 50,
                                ),
                              ),
                              const SizedBox(width: 10),
                              Text(
                                "${provider.weatherModel!.mainModels!.temp}°C",
                                style: const TextStyle(
                                  fontSize: 34,
                                  fontWeight: FontWeight.bold,
                                  color: Colors.white,
                                  shadows: [
                                    Shadow(
                                      blurRadius: 6.0,
                                      color: Colors.black26,
                                      offset: Offset(2.0, 2.0),
                                    ),
                                  ],
                                ),
                              ),
                            ],
                          ),
                          const SizedBox(height: 14.0),
                          Card(
                            elevation: 10,
                            shadowColor: Colors.blueGrey.withOpacity(0.3),
                            margin: const EdgeInsets.all(15),
                            shape: RoundedRectangleBorder(
                              borderRadius: BorderRadius.circular(20.0),
                            ),
                            child: Container(
                              decoration: BoxDecoration(
                                gradient: LinearGradient(
                                  colors: [
                                    Colors.blue.shade200,
                                    Colors.white,
                                  ],
                                  begin: Alignment.topLeft,
                                  end: Alignment.bottomRight,
                                ),
                                borderRadius: BorderRadius.circular(20.0),
                              ),
                              child: Padding(
                                padding: const EdgeInsets.all(20.0),
                                child: Column(
                                  crossAxisAlignment: CrossAxisAlignment.start,
                                  children: [
                                    Column(
                                      crossAxisAlignment:
                                          CrossAxisAlignment.start,
                                      children: [
                                        buildWeatherRow(
                                          icon: Icons.water_drop,
                                          title: "Humidity",
                                          value:
                                              "${provider.weatherModel!.mainModels!.humidity}%",
                                        ),
                                        buildWeatherRow(
                                          icon: Icons.air,
                                          title: "Wind Speed",
                                          value:
                                              "${provider.weatherModel!.windModel!.speed} m/s",
                                        ),
                                        buildWeatherRow(
                                          icon: Icons.compress,
                                          title: "Pressure",
                                          value:
                                              "${provider.weatherModel!.mainModels!.pressure} hPa",
                                        ),
                                        buildWeatherRow(
                                          icon: Icons.thermostat_outlined,
                                          title: "Feels Like",
                                          value:
                                              "${provider.weatherModel!.mainModels!.feelsLike}°C",
                                        ),
                                        buildWeatherRow(
                                          icon: Icons.home_mini,
                                          title: "Min Temperature",
                                          value:
                                              "${provider.weatherModel!.mainModels!.tempMin}°C",
                                        ),
                                        buildWeatherRow(
                                          icon: Icons.home_max,
                                          title: "Max Temperature",
                                          value:
                                              "${provider.weatherModel!.mainModels!.tempMax}°C",
                                        ),
                                        buildWeatherRow(
                                          icon: Icons.water,
                                          title: "Sea Level",
                                          value:
                                              "${provider.weatherModel!.mainModels!.seaLevel} m",
                                        ),
                                        buildWeatherRow(
                                          icon: Icons.landscape,
                                          title: "Ground Level",
                                          value:
                                              "${provider.weatherModel!.mainModels!.grndLevel} m",
                                        ),
                                      ],
                                    ),
                                    const Divider(color: Colors.black38),
                                    Row(
                                      mainAxisAlignment:
                                          MainAxisAlignment.spaceBetween,
                                      children: [
                                        Column(
                                          crossAxisAlignment:
                                              CrossAxisAlignment.start,
                                          children: [
                                            const Text(
                                              'Sunrise',
                                              style: TextStyle(
                                                  fontSize: 18,
                                                  fontWeight: FontWeight.bold,
                                                  color: Colors.black38),
                                            ),
                                            const SizedBox(height: 5),
                                            Container(
                                              decoration: BoxDecoration(
                                                color: Colors.orangeAccent
                                                    .withOpacity(0.3),
                                                borderRadius:
                                                    BorderRadius.circular(10),
                                                boxShadow: [
                                                  BoxShadow(
                                                    color: Colors.orangeAccent
                                                        .withOpacity(0.5),
                                                    blurRadius: 8,
                                                    spreadRadius: 2,
                                                    offset: const Offset(2, 2),
                                                  ),
                                                ],
                                              ),
                                              padding:
                                                  const EdgeInsets.symmetric(
                                                      vertical: 5,
                                                      horizontal: 15),
                                              child: Row(
                                                children: [
                                                  const Icon(
                                                    Icons.wb_sunny,
                                                    color: Colors.yellow,
                                                    size: 24,
                                                  ),
                                                  const SizedBox(width: 8),
                                                  Text(
                                                    formatTime(provider
                                                        .weatherModel!
                                                        .sysModel!
                                                        .sunrise),
                                                    style: const TextStyle(
                                                        fontSize: 20,
                                                        color: Colors.white,
                                                        fontWeight:
                                                            FontWeight.w500),
                                                  ),
                                                ],
                                              ),
                                            ),
                                          ],
                                        ),
                                        Column(
                                          crossAxisAlignment:
                                              CrossAxisAlignment.end,
                                          children: [
                                            const Text(
                                              'Sunset',
                                              style: TextStyle(
                                                  fontSize: 18,
                                                  fontWeight: FontWeight.bold,
                                                  color: Colors.black38),
                                            ),
                                            const SizedBox(height: 5),
                                            Container(
                                              decoration: BoxDecoration(
                                                color: Colors.deepPurpleAccent
                                                    .withOpacity(0.3),
                                                borderRadius:
                                                    BorderRadius.circular(10),
                                                boxShadow: [
                                                  BoxShadow(
                                                    color: Colors
                                                        .deepPurpleAccent
                                                        .withOpacity(0.5),
                                                    blurRadius: 8,
                                                    spreadRadius: 2,
                                                    offset: const Offset(2, 2),
                                                  ),
                                                ],
                                              ),
                                              padding:
                                                  const EdgeInsets.symmetric(
                                                      vertical: 5,
                                                      horizontal: 15),
                                              child: Row(
                                                children: [
                                                  const Icon(
                                                    Icons.nightlight_round,
                                                    color: Colors.white,
                                                    size: 24,
                                                  ),
                                                  const SizedBox(width: 8),
                                                  Text(
                                                    formatTime(provider
                                                        .weatherModel!
                                                        .sysModel!
                                                        .sunset),
                                                    style: const TextStyle(
                                                        fontSize: 20,
                                                        color: Colors.white,
                                                        fontWeight:
                                                            FontWeight.w500),
                                                  ),
                                                ],
                                              ),
                                            ),
                                          ],
                                        ),
                                      ],
                                    ),
                                  ],
                                ),
                              ),
                            ),
                          ),
                        ],
                      ),
                  ],
                ),
              ),
            ),
          );
        },
      ),
    );
  }

  Widget buildWeatherRow({
    required IconData icon,
    required String title,
    required String value,
  }) {
    return Padding(
      padding: const EdgeInsets.symmetric(vertical: 8.0),
      child: Row(
        children: [
          Icon(
            icon,
            color: Colors.blue.shade400,
            size: 24,
          ),
          const SizedBox(width: 10),
          Text(
            title,
            style: const TextStyle(
              fontSize: 18,
              fontWeight: FontWeight.bold,
              color: Colors.black87,
            ),
          ),
          const Spacer(),
          Text(
            value,
            style: const TextStyle(
              fontSize: 18,
              fontWeight: FontWeight.w500,
              color: Colors.black54,
            ),
          ),
        ],
      ),
    );
  }

  String formatTime(int? time) {
    if (time == null) return "";
    DateTime dateTime = DateTime.fromMillisecondsSinceEpoch(time * 1000);
    return "${dateTime.hour}:${dateTime.minute.toString().padLeft(2, '0')}";
  }
}

File: lib/helper/HomeScreen.java
import 'package:flutter/material.dart';
import 'package:provider/provider.dart';
import 'package:weather_app/provider/home_provider.dart';

class HomeScreen extends StatefulWidget {
  const HomeScreen({super.key});

  @override
  State<HomeScreen> createState() => _HomeScreenState();
}

class _HomeScreenState extends State<HomeScreen> {
  TextEditingController searchController = TextEditingController();

  String getBackgroundImage(String iconCode) {
    switch (iconCode) {
      case '01d':
        return 'assets/images/01d.jpeg';
      case '01n':
        return 'assets/images/01n.jpeg';
      case '02d':
        return 'assets/images/02d.jpeg';
      case '02n':
        return 'assets/images/02n.jpeg';
      case '03d':
        return 'assets/images/03d.jpeg';
      case '03n':
        return 'assets/images/03n.jpeg';
      case '04d':
        return 'assets/images/04d.jpeg';
      case '04n':
        return 'assets/images/04n.jpeg';
      case '09d':
        return 'assets/images/09d.jpeg';
      case '09n':
        return 'assets/images/09n.jpeg';
      case '10d':
        return 'assets/images/10d.jpeg';
      case '10n':
        return 'assets/images/10n.jpeg';
      case '11d':
        return 'assets/images/11d.jpeg';
      case '11n':
        return 'assets/images/11n.jpeg';
      case '13d':
        return 'assets/images/13d.jpeg';
      case '13n':
        return 'assets/images/13n.jpeg';
      case '50d':
        return 'assets/images/50d.jpeg';
      case '50n':
        return 'assets/images/50n.jpeg';
      default:
        return 'assets/default.jpeg';
    }
  }

  String getWeatherIcon(String iconCode) {
    switch (iconCode) {
      case '01d':
        return 'assets/icons/01d.png';
      case '01n':
        return 'assets/icons/01n.png';
      case '02d':
        return 'assets/icons/02d.png';
      case '02n':
        return 'assets/icons/02n.png';
      case '03d':
        return 'assets/icons/03d.png';
      case '03n':
        return 'assets/icons/03n.png';
      case '04d':
        return 'assets/icons/04d.png';
      case '04n':
        return 'assets/icons/04n.png';
      case '09d':
        return 'assets/icons/09d.png';
      case '09n':
        return 'assets/icons/09n.png';
      case '10d':
        return 'assets/icons/10d.png';
      case '10n':
        return 'assets/icons/10n.png';
      case '11d':
        return 'assets/icons/11d.png';
      case '11n':
        return 'assets/icons/11n.png';
      case '13d':
        return 'assets/icons/13d.png';
      case '13n':
        return 'assets/icons/13n.png';
      case '50d':
        return 'assets/icons/50d.png';
      case '50n':
        return 'assets/icons/50n.png';
      default:
        return '🔅';
    }
  }

  @override
  void initState() {
    super.initState();
    context.read<HomeProvider>().loadBookmarkedCities();
  }

  void searchCityWeather() {
    String city = searchController.text.trim();
    if (city.isNotEmpty) {
      context.read<HomeProvider>().fetchWeatherData(city: city);
    }
  }

  void addToBookmark(String city) {
    context.read<HomeProvider>().bookmarkCity(city).then((_) {
      ScaffoldMessenger.of(context).showSnackBar(
        SnackBar(content: Text('$city has been added to bookmarks')),
      );
    });
  }

  String formatTime(int? time) {
    if (time == null) return "N/A";
    DateTime dateTime = DateTime.fromMillisecondsSinceEpoch(time * 1000);
    return "${dateTime.hour}:${dateTime.minute.toString().padLeft(2, '0')}";
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      body: Consumer<HomeProvider>(
        builder: (context, provider, child) {
          if (provider.weatherModel == null) {
            return const Center(child: CircularProgressIndicator());
          }

          String backgroundImage =
              getBackgroundImage(provider.weatherModel!.weathersList![0].icon!);

          return Container(
            decoration: BoxDecoration(
              gradient: LinearGradient(
                begin: Alignment.topLeft,
                end: Alignment.bottomRight,
                colors: [
                  Colors.blue.withOpacity(0.7),
                  Colors.white.withOpacity(0.5)
                ],
              ),
              image: DecorationImage(
                image: AssetImage(backgroundImage),
                fit: BoxFit.cover,
                opacity: 0.5,
              ),
            ),
            child: Padding(
              padding: const EdgeInsets.all(16.0),
              child: Column(
                crossAxisAlignment: CrossAxisAlignment.start,
                children: [
                  // Search Field
                  Padding(
                    padding: const EdgeInsets.only(top: 20.0),
                    child: TextField(
                      controller: searchController,
                      decoration: InputDecoration(
                        hintText: "Search city...",
                        suffixIcon: IconButton(
                          icon: const Icon(Icons.search),
                          onPressed: searchCityWeather,
                        ),
                        filled: true,
                        fillColor: Colors.white.withOpacity(0.8),
                        border: OutlineInputBorder(
                          borderRadius: BorderRadius.circular(30),
                          borderSide: BorderSide.none,
                        ),
                      ),
                      onSubmitted: (value) => searchCityWeather(),
                    ),
                  ),
                  const SizedBox(height: 16),
                  // Weather Data
                  if (provider.weatherModel != null)
                    Expanded(
                      child: Column(
                        mainAxisAlignment: MainAxisAlignment.center,
                        children: [
                          Text(
                            provider.weatherModel!.name ?? "",
                            style: const TextStyle(
                              fontSize: 32,
                              fontWeight: FontWeight.bold,
                              color: Colors.white,
                            ),
                          ),
                          const SizedBox(height: 16),
                          Image.asset(
                            getWeatherIcon(
                                provider.weatherModel!.weathersList![0].icon!),
                            width: 100,
                          ),
                          Text(
                            "${provider.weatherModel!.mainModels!.temp}°C",
                            style: const TextStyle(
                              fontSize: 24,
                              color: Colors.white,
                            ),
                          ),
                        ],
                      ),
                    ),
                ],
              ),
            ),
          );
        },
      ),
    );
  }
}
